yb-voyager/src/version/yb_version.go

```diff
@@ -0,0 +1,150 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package version
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/hashicorp/go-version"
+	"golang.org/x/exp/slices"
+)
+
+var supportedYBVersionStableSeriesOld = []string{"2.20"}
+var supportedYBVersionStableSeries = []string{"2024.1"}
+var supportedYBVersionPreviewSeries = []string{"2.21"}
+
+var allSupportedYBVersionSeries = append(append(supportedYBVersionStableSeries, supportedYBVersionPreviewSeries...), supportedYBVersionStableSeriesOld...)
+
+const (
+	STABLE     = "stable"
+	PREVIEW    = "preview"
+	STABLE_OLD = "stable_old"
+)
+
+/*
+YBVersion is a wrapper around hashicorp/go-version.Version that adds some Yugabyte-specific
+functionality.
+ 1. It only supports versions with 2, 3, or 4 segments (A.B, A.B.C, A.B.C.D).
+ 2. It only accepts one of supported Yugabyte version series.
+ 3. It provides a method to compare versions based on common prefix of segments. This is useful in cases
+    where the version is not fully specified (e.g. 2.20 vs 2.20.7.0).
+    If the user specifies 2.20, a reasonable assumption is that they would be on the latest 2.20.x.y version.
+    Therefore, we want only want to compare the prefix 2.20 between versions and ignore the rest.
+*/
+type YBVersion struct {
+	*version.Version
+}
+
+func NewYBVersion(v string) (*YBVersion, error) {
+	v1, err := version.NewVersion(v)
+	if err != nil {
+		return nil, err
+	}
+
+	ybv := &YBVersion{v1}
+	origSegLen := ybv.originalSegmentsLen()
+	if origSegLen < 2 || origSegLen > 4 {
+		return nil, fmt.Errorf("invalid YB version: %s. Version should have between min 2 and max 4 segments (A.B.C.D). Version %s has ybv.originalSegmentsLen()", v, v)
+	}
+
+	if !slices.Contains(allSupportedYBVersionSeries, ybv.Series()) {
+		return nil, fmt.Errorf("unsupported YB version series: %s. Supported YB version series = %v", ybv.Series(), allSupportedYBVersionSeries)
+	}
+	return ybv, nil
+}
+
+func (ybv *YBVersion) Series() string {
+	return joinIntsWith(ybv.Segments()[:2], ".")
+}
+
+func (ybv *YBVersion) ReleaseType() string {
+	series := ybv.Series()
+	if slices.Contains(supportedYBVersionStableSeries, series) {
+		return STABLE
+	} else if slices.Contains(supportedYBVersionPreviewSeries, series) {
+		return PREVIEW
+	} else if slices.Contains(supportedYBVersionStableSeriesOld, series) {
+		return STABLE_OLD
+	} else {
+		panic("unknown release type for series: " + series)
+	}
+}
+
+func (ybv *YBVersion) originalSegmentsLen() int {
+	orig := ybv.Original()
+	segments := strings.Split(orig, ".")
+	return len(segments)
```

Comment:
why not use version.Segments() here instaed of splitting?

Reply:
Added comment here as well. Essentially: 

This returns the len of the segments in the original input. For instance if input is 2024.1,
- go-version.Version.Segments() will return [2024, 1, 0, 0] (assume 2024.1.0.0)
- original = 2024.1
- originalSegmentsLen  = 2 ([2024,1])

Reply:
got it